src/yeoda/datacube.py

```diff
@@ -1182,540 +644,343 @@ def align_dimension(self, dc_other, name, inplace=False):
+        super().__init__(writer)
 
-    sdim_name = dcs[0].sdim_name
-    tdim_name = dcs[0].tdim_name
-    if sdim_name not in list(intersected_inventory.keys()):
-        sdim_name = None
+    @classmethod
+    def from_data(cls, data, dirpath, filename_class=None, stack_groups=None, fn_groups_map=None, def_fields=None,
+                  format='.nc', mosaic=None, stack_dimension='layer_id', tile_dimension='tile_id', **kwargs) -> "DataCubeWriter":
+
+        tile_ids = ['0'] if mosaic is None else mosaic.all_tile_names
+        stack_ids = data[stack_dimension]
+        filepaths = cls._get_filepaths_from_tile_stack_ids(tile_ids, stack_ids, filename_class, dirpath,
+                                                           tile_dimension=tile_dimension,
+                                                           stack_dimension=stack_dimension,
+                                                           stack_groups=stack_groups,
+                                                           fn_groups_map=fn_groups_map,
+                                                           def_fields=def_fields)
+
+        fr_dict = {'filepath': filepaths,
+                   stack_dimension: stack_ids,
+                   tile_dimension: tile_ids}
+        file_register = pd.DataFrame(fr_dict)
+        writer_class = RASTER_DATA_CLASS[format][1]
+        return writer_class.from_xarray(data, file_register, mosaic=mosaic, **kwargs)
+
+    @staticmethod
+    def _get_filepaths_from_tile_stack_ids(tile_ids, stack_ids, filename_class, dirpath, tile_dimension='tile_id',
+                                           stack_dimension='layer_id', stack_groups=None, fn_groups_map=None,
+                                           def_fields=None):
+        fn_groups_map = {} if fn_groups_map is None else fn_groups_map
+        def_fields = {} if def_fields is None else def_fields
+        if filename_class is None:
+            fields_def = dict([
+                (stack_dimension, {}),
+                (tile_dimension, {})])
+            filename_class = create_fn_class(fields_def)
+
+        tile_ids_aligned = []
+        filepaths = []
+        for t, tile_id in enumerate(tile_ids):
+            for s, stack_id in enumerate(stack_ids):
+                fields = dict()
+                fields[tile_dimension] = tile_id
+                fields.update(def_fields)
+                if stack_groups is not None:
+                    group_id = stack_groups[s]
+                    fields.update({stack_dimension: group_id})
+                    fields.update(fn_groups_map.get(group_id))
+                else:
+                    fields.update({stack_dimension: stack_id})
 
-    if tdim_name not in list(intersected_inventory.keys()):
-        tdim_name = None
+                filename = str(filename_class(fields, ext=format))
+                filepaths.append(os.path.join(dirpath, filename))
+                tile_ids_aligned.append(tile_id)
```

Comment:
looks to me like this could be extracted into a function with a nice name

Reply:
I kept it as is for the time being, but yes, might be a good idea.